include/dd/Package.hpp

```diff
@@ -1232,20 +1308,57 @@ namespace dd {
                             e2 = yCopy;
                         }
 
-                        auto m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start);
-
-                        if (k == 0 || edge[idx].w == Complex::zero) {
-                            edge[idx] = m;
-                        } else if (m.w != Complex::zero) {
-                            auto old_e = edge[idx];
-                            edge[idx]  = add2(edge[idx], m);
-                            cn.returnToCache(old_e.w);
-                            cn.returnToCache(m.w);
+                        if constexpr (std::is_same_v<LeftOperandNode, dNode>) {
+                            dEdge m;
+                            dEdge::applyDmChangesToEdges(&e1, &e2);
+                            if (!generateDensityMatrix || idx == 1) {
+                                // When generateDensityMatrix is false or I have the first edge I don't optimize anything and set generateDensityMatrix for all child edges
+                                m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start, false);
+                            } else if (idx == 2) {
+                                // When I have the second edge and generateDensityMatrix == false, then edge[2] == edge[1]
+                                if (k == 0) {
+                                    if (edge[1].w.approximatelyZero()) {
+                                        edge[2] = ResultEdge::zero;
+                                    } else {
+                                        edge[2] = ResultEdge{edge[1].p, cn.getCached(edge[1].w.r->value, edge[1].w.i->value)};
+                                    }
+                                }
+                                continue;
+                            } else {
+                                m = multiply2(e1, e2, static_cast<Qubit>(var - 1), start, generateDensityMatrix);
+                            }
+                            if (k == 0 || edge[idx].w == Complex::zero) {
+                                edge[idx] = m;
+                            } else if (m.w != Complex::zero) {
```

Comment:
Here I need to compare against Complex::zero specifically, otherwise, this leads to a caching error.

Reply:
What?! 🙃 the certainly implies some underlying bug.
`e.w != Complex::zero` resolves to `e.w.r != &ComplexTable<>::zero && e.w.i != &ComplexTable<>::zero` which is precisely what the `exactlyZero` function does.

I would suspect, that this might again be due to the `ComplexTable<>` here.